src/lib/externaldata.py

```diff
@@ -454,3 +454,51 @@ async def validate_checker_data(self, external_data: ExternalBase):
 
     async def check(self, external_data: ExternalBase):
         raise NotImplementedError
+
+    # Various helplers for checkers; assumed to be safely usable only from subclasses
+
+    async def _complete_digests(
+        self, url: t.Union[str, URL], digests: MultiDigest
+    ) -> MultiDigest:
+        """
+        Re-download given `url`, verify it against given `digests`,
+        and return a `MultiDigest` with all digest types set.
+        """
+        multihash = MultiHash()
+        try:
+            async with self.session.get(url) as resp:
+                async for chunk, _ in resp.content.iter_chunks():
+                    multihash.update(chunk)
+        except NETWORK_ERRORS as err:
+            raise CheckerFetchError from err
+        new_digests = multihash.hexdigest()
+        if new_digests != digests:
+            raise CheckerFetchError(
+                f"Checksum mismatch for {url}: "
+                f"expected {digests}, got {new_digests}"
+            )
+        return new_digests
+
+    async def _set_new_version(self, source: ExternalBase, new_version: ExternalState):
+        """
+        Set the `new_version` for `source`, ensuring common digest types are set.
+        """
+        if (
+            isinstance(source, ExternalData) and isinstance(new_version, ExternalFile)
+        ) and not (
+            source.current_version.checksum.digests & new_version.checksum.digests
+        ):
+            log.warning(
+                "Source %s: %s didn't get a suitable checksum type for this source; "
+                "consider replacing %s with %s in the manifest",
```

Comment:
I believe but could be wrong that extra-data sources can only have one checkum type, sha256 iirc. I think this advice is only applicable in the case where a normal source is being checked?

Reply:
Indeed, extra-data only supports sha256, so this advice may be misleading. Should we change the advice to mention the extra-data limitation, or just not give for extra-data?

Reply:
I think it would be better to not give the advice at all for extra-data sources.